src/protocolsupport/protocol/packet/middleimpl/serverbound/play/v_pe/PositionLook.java

```diff
@@ -26,47 +34,40 @@ public void readFromClientData(ByteBuf clientdata) {
 		y = clientdata.readFloatLE() - 1.6200000047683716D;
 		z = clientdata.readFloatLE();
 		pitch = clientdata.readFloatLE();
+		headYaw = clientdata.readFloatLE();
 		yaw = clientdata.readFloatLE();
-		clientdata.readFloatLE(); //head yaw
-		clientdata.readByte(); //mode
+		mode = clientdata.readByte(); //mode
 		onGround = clientdata.readBoolean();
-		VarNumberSerializer.readVarLong(clientdata);
+		vehicle = VarNumberSerializer.readVarLong(clientdata);
+		if (mode == 2) {
+			huh1 = VarNumberSerializer.readSVarInt(clientdata);
+			huh2 = VarNumberSerializer.readSVarInt(clientdata);
+		}
 	}
 
 	@Override
 	public RecyclableCollection<ServerBoundPacketData> toNative() {
-		MovementCache movecache = cache.getMovementCache();
 		RecyclableArrayList<ServerBoundPacketData> packets = RecyclableArrayList.create();
+		MovementCache movecache = cache.getMovementCache();
+		NetworkEntity player = cache.getWatchedEntityCache().getSelfPlayer();
 		movecache.updatePEPositionLeniency(y);
 		movecache.setPEClientPosition(x, y, z);
 		//PE doesn't send a movement confirm after position set, so we just confirm teleport straight away
 		int teleportId = movecache.teleportConfirm();
 		if (teleportId != -1) {
 			packets.add(MiddleTeleportAccept.create(teleportId));
-			packets.add(MiddlePositionLook.create(movecache.getX(), movecache.getY(), movecache.getZ(), yaw, pitch, onGround));
+			packets.add(MiddlePositionLook.create(movecache.getX(), movecache.getY(), movecache.getZ(), headYaw, pitch, onGround));
+			//TODO: Play around more with these numbers to perhaps make things even more smooth.
+			x = Math.floor(x * 8) / 8; y = Math.ceil((y + 0.3) * 8) / 8; z = Math.floor(z * 8) / 8;
```

Comment:
Is there a point in this, because i hasn't seen having any use in rounding values.

Reply:
Maybe the chances of PE sending a correct movement packet will increase and the teleporting back will decrease. However I haven't done enough testing on that so I will remove it for now.